controller/cmd/service-mirror/probe_manager.go

```diff
@@ -0,0 +1,198 @@
+	return &ProbeManager{
+		k8sAPI:       k8sAPI,
+		probeWorkers: make(map[string]*ProbeWorker),
+		events:       events,
+		metricVecs:   &metricVecs,
+		done:         make(chan struct{}, 1),
+	}
+}
+
+// ProbeManager takes care of managing the lifecycle of probe workers
+type ProbeManager struct {
+	k8sAPI       *k8s.API
+	probeWorkers map[string]*ProbeWorker
+
+	events     chan interface{}
+	metricVecs *probeMetricVecs
+	done       chan struct{}
+}
+
+// GatewayProbeSpec is the specification of a probe that will be executed by the worker
+type GatewayProbeSpec struct {
+	clusterName   string
+	gatewayName   string
+	gatewayNs     string
+	gatewayIps    []string
+	port          int32
+	path          string
+	periodSeconds int32
+}
+
+// MirroredServiceUnpaired is emitted when a service is no longer mirrored
+type MirroredServiceUnpaired struct {
+	serviceName      string
+	serviceNamespace string
+	gatewayName      string
+	gatewayNs        string
+	clusterName      string
+}
+
+// MirroredServicePaired is emitted when a new mirrored service is created
+type MirroredServicePaired struct {
+	serviceName      string
+	serviceNamespace string
+	*GatewayProbeSpec
+}
+
+// ClusterRegistered is emitted when a new cluster becomes registred for mirroring
+type ClusterRegistered struct {
+	clusterName   string
+	port          int32
+	path          string
+	periodSeconds int32
+}
+
+// GatewayUpdated is emitted when something about the gateway is updated (i.e. its external IP(s))
+type GatewayUpdated struct {
+	*GatewayProbeSpec
+}
+
+func probeKey(gatewayNamespace string, gatewayName string, clusterName string) string {
+	return fmt.Sprintf("%s-%s-%s", gatewayNamespace, gatewayName, clusterName)
+}
+
+func (m *ProbeManager) handleEvent(ev interface{}) {
+	switch ev := ev.(type) {
+	case *MirroredServicePaired:
+		m.handleMirroredServicePaired(ev)
+	case *MirroredServiceUnpaired:
+		m.handleMirroredServiceUnpaired(ev)
+	case *GatewayUpdated:
+		m.handleGatewayUpdated(ev)
+	case *ClusterRegistered:
+		m.handleClusterRegistered(ev)
+	}
+}
+
+func (m *ProbeManager) handleMirroredServicePaired(event *MirroredServicePaired) {
+	probeKey := probeKey(event.gatewayNs, event.gatewayName, event.clusterName)
+	worker, ok := m.probeWorkers[probeKey]
+	if ok {
+		log.Debugf("Probe worker %s already exists", probeKey)
+		worker.IncNumServices()
+	} else {
+		log.Debugf("Creating probe worker %s", probeKey)
+		probeMetrics := m.metricVecs.newMetrics(event.gatewayNs, event.gatewayName, event.clusterName)
+		worker = NewProbeWorker(event.GatewayProbeSpec, &probeMetrics, probeKey)
+		worker.IncNumServices()
+		m.probeWorkers[probeKey] = worker
+		worker.Start()
+	}
+}
+
+func (m *ProbeManager) handleMirroredServiceUnpaired(event *MirroredServiceUnpaired) {
+	probeKey := probeKey(event.gatewayNs, event.gatewayName, event.clusterName)
+	worker, ok := m.probeWorkers[probeKey]
+	if ok {
+		worker.DcrNumServices()
+		if worker.numAssociatedServices < 1 {
```

Comment:
The assumption is that we cannot. But we might have a bug somewhere that allows that. What do you suggest? Keep a set of associated services in the probe_worker and only decrement if this service that is getting unpaired is in there?

Reply:
I guess my question is how confident are we about this assumption?  At the very least, we should document that the contract here is that we must not get duplicate or out-of-order events.

If you trace it all the way back, these events are triggered by Kubernetes informer events.  Do informers make any guarantees about duplicate or out-of-order events?  Or does our handling of those events guarantee that events will be properly de-duped by the time they make it here?

Reply:
@adleong very good question! So yes, all the way back these events are triggered by the informer. And indeed you are right that we can get duplicates here. However, if you trace a bit forward these particular events are triggered by the cluster_watcher. The cluster watcher assumes that you might get duplicate events from the informers (i.e. if your watcher fails and you fall back to perioding listing). because of that it makes sure that we do not create or delete a service twice. So for example if you get a service created event for a service that you already have and that has not changed, it will not try to crete the service again and as a result it will not issue an event to the prober twice. Will look at it again, but as far as I remember, this is how it was designed. I will put a comment though. I think all this deserves a good dose of tests to verify that all these edge cases are handled correctly (which I will write in time). 

That being said, in my latest commit I transitioned to keeping a set of these services (well a map, as Go does not have sets...) to track the count. That alone should serve as a second gatekeeper against duplication. Does that sound correct ?